File: isBST.java
//
// Binary trees are already defined with this interface:
// class Tree<T> {
//   Tree(T x) {
//     value = x;
//   }
//   T value;
//   Tree<T> left;
//   Tree<T> right;
// }
List<Integer> l = new ArrayList();

void makeList(Tree<Integer> t) {
  if (t != null) {
    makeList(t.left);
    l.add(t.value);
    makeList(t.right);
  }
}

boolean isBST(Tree<Integer> t) {
  makeList(t);
  
  for (int i = 1; i < l.size(); i++) {
    if (l.get(i - 1) >= l.get(i)) {
      return false;
    }
  }
  
  return true;
}

File: python/bool/isBST.py
# https://afteracademy.com/blog/check-if-a-binary-tree-is-BST-or-not
# Binary trees are already defined with this interface:
# class Tree(object):
#   def __init__(self, x):
#     self.value = x
#     self.left = None
#     self.right = None
def helper(root, range_min, range_max):
    if(root == None):
        return True
    
    if (root.value < range_min or root.value > range_max):
        return False
    
    if (helper(root.left, range_min, root.value - 1)):
        if (helper(root.right, root.value + 1, range_max)):
            return True
    return False

INT_MIN = -2147483648
INT_MAX = 2147483647

def isBST(root):
    if (helper(root, INT_MIN, INT_MAX)):
        return True
    return False
    
